File: src/services/api/smartAlbums.ts
/**
 * 智能相册 API
 */

import { invoke } from '@tauri-apps/api/core';
import type { SearchFilters } from '@/types';

/**
 * 智能相册
 */
export interface SmartAlbum {
  smartAlbumId: number;
  name: string;
  description?: string;
  filters: SearchFilters;
  icon?: string;
  color?: string;
  dateCreated: string;
  dateModified: string;
  sortOrder: number;
}

/**
 * 创建智能相册参数
 */
export interface CreateSmartAlbum {
  name: string;
  description?: string;
  filters: SearchFilters;
  icon?: string;
  color?: string;
}

/**
 * 更新智能相册参数
 */
export interface UpdateSmartAlbum {
  name?: string;
  description?: string;
  filters?: SearchFilters;
  icon?: string;
  color?: string;
  sortOrder?: number;
}

/**
 * 创建智能相册
 */
export async function createSmartAlbum(album: CreateSmartAlbum): Promise<number> {
  return invoke<number>('create_smart_album', { album });
}

/**
 * 获取智能相册
 */
export async function getSmartAlbum(smartAlbumId: number): Promise<SmartAlbum | null> {
  return invoke<SmartAlbum | null>('get_smart_album', { smartAlbumId });
}

/**
 * 获取所有智能相册
 */
export async function getAllSmartAlbums(): Promise<SmartAlbum[]> {
  return invoke<SmartAlbum[]>('get_all_smart_albums');
}

/**
 * 更新智能相册
 */
export async function updateSmartAlbum(
  smartAlbumId: number,
  update: UpdateSmartAlbum
): Promise<boolean> {
  return invoke<boolean>('update_smart_album', { smartAlbumId, update });
}

/**
 * 删除智能相册
 */
export async function deleteSmartAlbum(smartAlbumId: number): Promise<boolean> {
  return invoke<boolean>('delete_smart_album', { smartAlbumId });
}

File: src-tauri/src/commands/smart_albums.rs
//! 智能相册相关命令

use tauri::State;

use crate::db::{SmartAlbum, CreateSmartAlbum, UpdateSmartAlbum};
use crate::AppState;

/// 创建智能相册
#[tauri::command]
pub async fn create_smart_album(
    state: State<'_, AppState>,
    album: CreateSmartAlbum,
) -> Result<i64, String> {
    state
        .db
        .create_smart_album(&album)
        .map_err(|e| e.to_string())
}

/// 获取智能相册
#[tauri::command]
pub async fn get_smart_album(
    state: State<'_, AppState>,
    smart_album_id: i64,
) -> Result<Option<SmartAlbum>, String> {
    state
        .db
        .get_smart_album(smart_album_id)
        .map_err(|e| e.to_string())
}

/// 获取所有智能相册
#[tauri::command]
pub async fn get_all_smart_albums(
    state: State<'_, AppState>,
) -> Result<Vec<SmartAlbum>, String> {
    state
        .db
        .get_all_smart_albums()
        .map_err(|e| e.to_string())
}

/// 更新智能相册
#[tauri::command]
pub async fn update_smart_album(
    state: State<'_, AppState>,
    smart_album_id: i64,
    update: UpdateSmartAlbum,
) -> Result<bool, String> {
    state
        .db
        .update_smart_album(smart_album_id, &update)
        .map_err(|e| e.to_string())
}

/// 删除智能相册
#[tauri::command]
pub async fn delete_smart_album(
    state: State<'_, AppState>,
    smart_album_id: i64,
) -> Result<bool, String> {
    state
        .db
        .delete_smart_album(smart_album_id)
        .map_err(|e| e.to_string())
}
